test/dynamo/test_package.py

```diff
@@ -0,0 +1,185 @@
+# Owner(s): ["module: dynamo"]
+
+import os
+import pickle
+
+import torch
+import torch._dynamo.testing
+import torch._inductor.config
+import torch._inductor.test_case
+import torch.onnx.operators
+import torch.utils.cpp_extension
+from torch._dynamo.package import _CompilePackage
+from torch._inductor.runtime.runtime_utils import cache_dir
+
+
+class TestStorageContext:
+    def __init__(self, path: str):
+        self.path = path
+        self.backends = {}
+
+    def _write_pickle(self, data, *path: str):
+        with open(os.path.join(self.path, *path) + ".pickle", "wb") as f:
+            pickle.dump(data, f)
+
+    def write_dynamo(self, dynamo):
+        self._write_pickle(dynamo, "dynamo")
+
+    def write_backend(self, backend_id):
+        os.makedirs(os.path.join(self.path, backend_id), exist_ok=True)
+        self._write_pickle(self.backends[backend_id], backend_id, "fx_graph")
+
+    def _read_pickle(self, *path):
+        with open(os.path.join(self.path, *path) + ".pickle", "rb") as f:
+            return pickle.load(f)
+
+    def read_backend(self, backend_id):
+        return self._read_pickle(backend_id, "fx_graph")
+
+    def read_dynamo(self):
+        return self._read_pickle("dynamo")
+
+    def add_backend(self, backend_id, backend):
+        self.backends[backend_id] = backend
+
+
+class TestPackage(torch._inductor.test_case.TestCase):
+    def context(self):
+        path = os.path.join(cache_dir(), f"package_{self.id()}")
+        os.makedirs(path, exist_ok=True)
+        return TestStorageContext(path)
+
+    def test_basic_fn(self):
+        ctx = self.context()
+
+        def fn(x):
+            return x + 1
+
+        args = (torch.randn(3, 2),)
+
+        # Saving
+        package = _CompilePackage(fn)
+        compiled_fn = torch._dynamo.optimize(backend="eager", package=package)(fn)
+        expected = compiled_fn(*args)
+        for backend_id, backend in package.cached_backends.items():
+            ctx.add_backend(backend_id, backend)
+        package.save(ctx)
+
+        # Loading
+        torch._dynamo.reset()
+        with torch.compiler.set_stance("fail_on_recompile"):
+            with self.assertRaisesRegex(
+                RuntimeError,
+                "Detected recompile when torch.compile stance is 'fail_on_recompile'",
+            ):
+                compiled_fn(*args)
+
+            package = _CompilePackage(fn, ctx.read_dynamo())
```

Comment:
Eventually, _CompilePackage is going to need to take in a list of backends, right, not just dynamo state? It says `cached_backends` is only used for testing, but doesn't there still need to be a mapping from a CompilePackage to a list of backend ids?

Reply:
@jamesjwu  Correct, right now we just store all the backend ids inside DynamoCodeCacheEntry and manually aggregate them when needed, so they are not explicitly stored in a dedicated list.